src/lib/snapper/snapshot_dbus.rb

```diff
@@ -0,0 +1,231 @@
+# encoding: utf-8
+
+# ------------------------------------------------------------------------------
+# Copyright (c) 2016 SUSE LLC. All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it under
+# the terms of version 2 of the GNU General Public License as published by the
+# Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along with
+# this program; if not, contact Novell, Inc.
+#
+# To contact Novell about this file by physical or electronic mail, you may find
+# current contact information at www.novell.com.
+# ------------------------------------------------------------------------------
+
+require "yast"
+require "dbus"
+
+module Yast
+  class SnapshotDBus
+    extend Forwardable
+    include Yast::Logger
+
+    SNAPSHOT_ID = {
+      0 => "single",
+      1 => "pre",
+      2 => "post"
+    }.freeze
+
+    def_delegators :@context, :config, :number, :pre_number
+
+    attr_accessor :context
+
+    def initialize(snapshot = nil)
+      @context = snapshot
+    end
+
+    def all(configs)
+      configs ||= list_configs
+      dbus_snapshots = Array(configs).each_with_object([]) do |config, snapshot_list|
+        list_snapshots(config).each do |dbus_snapshot|
+          next if dbus_snapshot[0] == 0
+          snapshot_list << {
+            type:        SNAPSHOT_ID[dbus_snapshot[1]],
+            number:      dbus_snapshot[0],
+            pre_number:  dbus_snapshot[2].to_i,
+            timestamp:   Time.at(dbus_snapshot[3]),
+            uid:         dbus_snapshot[4],
+            description: unescape(dbus_snapshot[5]),
+            cleanup:     unescape(dbus_snapshot[6]),
+            user_data:   unescape(dbus_snapshot[7]),
+            config:      config
+          }
+        end
+      end
+
+      dbus_snapshots
+    end
+
+    def create
+      attrs = [config || "", pre_number, description, cleanup, user_data]
+
+      attrs.delete_at(1) if !@context.post?
+      log.info "Creating snapshot #{@context.inspect} with result #{attrs.inspect}."
+
+      result = SnapshotDBus.dbus_object.send("Create#{@context.name.capitalize}Snapshot", *attrs).first
```

Comment:
It could be seen as the Snapshot Type, and Snapshot Dbus have different constructors to each of them, having pre and single snapshots the same attributes, and post having also the presnapshot_number.